experimental/harfbuzz/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+        core::ptr::write(ab, c as hb_codepoint_t);
+        true as hb_bool_t
+    } else {
+        false as hb_bool_t
+    }
+}
+
+unsafe extern "C" fn icu4x_hb_unicode_compose_destroy(user_data: *mut c_void) {
+    let _ = Box::from_raw(user_data as *mut CanonicalComposition);
+}
+
+unsafe extern "C" fn icu4x_hb_unicode_decompose(
+    _ufuncs: *mut hb_unicode_funcs_t,
+    ab: hb_codepoint_t,
+    a: *mut hb_codepoint_t,
+    b: *mut hb_codepoint_t,
+    user_data: *mut c_void,
+) -> hb_bool_t {
+    // Could we trust HarfBuzz to pass a valid scalar value?
+    // `hb_buffer_add_codepoints` seems to allow the introduction
+    // of non-scalar-value items to the HarfBuzz buffer. Should we
+    // treat that as conceptually `unsafe` so that if the app has
+    // introduced values that are not valid scalar values, we'd
+    // be OK with UB? OTOH, HarfBuzz docs say "code points", so
+    // following docs to the letter could still result in
+    // surrogate code points even if they don't make sense.
+    let composed = if let Some(composed) = core::char::from_u32(ab) {
+        composed
+    } else {
+        return false as hb_bool_t;
+    };
+    match (*(user_data as *mut CanonicalDecomposition)).decompose(composed) {
+        Decomposed::Default => false as hb_bool_t,
+        Decomposed::Expansion(first, second) => {
+            core::ptr::write(a, first as hb_codepoint_t);
+            core::ptr::write(b, second as hb_codepoint_t);
+            true as hb_bool_t
+        }
+        Decomposed::Singleton(single) => {
+            core::ptr::write(a, single as hb_codepoint_t);
+            core::ptr::write(b, 0);
+            true as hb_bool_t
+        }
+    }
+}
+
+unsafe extern "C" fn icu4x_hb_unicode_decompose_destroy(user_data: *mut c_void) {
+    let _ = Box::from_raw(user_data as *mut CanonicalDecomposition);
+}
+
+/// RAII holder for `*mut hb_unicode_funcs_t`.
+pub struct UnicodeFuncs {
+    raw: *mut hb_unicode_funcs_t,
+}
+
+impl UnicodeFuncs {
+    /// Takes ownership of a `*mut hb_unicode_funcs_t` without incrementing
+    /// the refcount.
+    ///
+    /// # Safety
+    ///
+    /// After the call, the previous owner must not call
+    /// `hb_unicode_funcs_destroy()`, since `UnicodeFuncs` will now
+    /// take care of it.
+    pub unsafe fn from_raw(raw: *mut hb_unicode_funcs_t) -> Self {
+        Self { raw }
+    }
+
+    /// Transfers the ownership of the wrapped pointer to the caller.
+    /// The caller is responsible for calling `hb_unicode_funcs_destroy()`;
+    /// `UnicodeFuncs` will no longer take care of it.
+    pub fn into_raw(funcs: Self) -> *mut hb_unicode_funcs_t {
+        let ret = funcs.raw;
+        core::mem::forget(funcs);
+        ret
+    }
+
+    /// Borrows the wrapped raw pointer without transferring ownership
+    /// and without affecting the refcount.
+    pub fn as_ptr(&self) -> *mut hb_unicode_funcs_t {
+        self.raw
+    }
```

Comment:
Note: I don't see where these functions are being called

Reply:
These are meant for users of this crate to potentially call.